crates/ide-diagnostics/src/handlers/trait_impl_redundant_assoc_item.rs

```diff
@@ -10,47 +17,196 @@ pub(crate) fn trait_impl_redundant_assoc_item(
-            TypeAlias::from(id)
-                .source(db)
-                .map(|it| it.syntax().value.text_range())
-                .unwrap_or(default_range),
-        ),
+    let (redundant_item_name, diagnostic_range, redundant_item_def) = match assoc_item {
+        hir::AssocItem::Function(id) => {
+            let function = Function::from(id);
+            (
+                format!("`fn {}`", redundant_assoc_item_name),
+                function
+                    .source(db)
+                    .map(|it| it.syntax().value.text_range())
+                    .unwrap_or(default_range),
+                format!("\n    {};", function.display(db).to_string()),
+            )
+        }
+        hir::AssocItem::Const(id) => {
+            let constant = Const::from(id);
+            (
+                format!("`const {}`", redundant_assoc_item_name),
+                constant
+                    .source(db)
+                    .map(|it| it.syntax().value.text_range())
+                    .unwrap_or(default_range),
+                format!("\n    {};", constant.display(db).to_string()),
+            )
+        }
+        hir::AssocItem::TypeAlias(id) => {
+            let type_alias = TypeAlias::from(id);
+            (
+                format!("`type {}`", redundant_assoc_item_name),
+                type_alias
+                    .source(db)
+                    .map(|it| it.syntax().value.text_range())
+                    .unwrap_or(default_range),
+                format!("\n    type {};", type_alias.name(ctx.sema.db).to_smol_str()),
+            )
+        }
     };
 
     Diagnostic::new(
         DiagnosticCode::RustcHardError("E0407"),
         format!("{redundant_item_name} is not a member of trait `{trait_name}`"),
         FileRange { file_id: d.file_id.file_id().unwrap(), range: diagnostic_range },
     )
+    .with_fixes(quickfix_for_redundant_assoc_item(
+        ctx,
+        d,
+        redundant_item_def,
+        diagnostic_range,
+    ))
+}
+
+/// add assoc item into the trait def body
+fn quickfix_for_redundant_assoc_item(
+    ctx: &DiagnosticsContext<'_>,
+    d: &hir::TraitImplRedundantAssocItems,
+    redundant_item_def: String,
+    range: TextRange,
+) -> Option<Vec<Assist>> {
+    let add_assoc_item_def = |builder: &mut SourceChangeBuilder| -> Option<()> {
+        let db = ctx.sema.db;
+        let root = db.parse_or_expand(d.file_id);
+        // don't modify trait def in outer crate
+        let current_crate = ctx.sema.scope(&d.impl_.syntax_node_ptr().to_node(&root))?.krate();
+        let trait_def_crate = d.trait_.module(db).krate();
+        if trait_def_crate != current_crate {
+            return None;
+        }
+        let trait_def = d.trait_.source(db)?.value;
+        let where_to_insert = trait_def
+            .syntax()
+            .descendants_with_tokens()
+            .find(|it| it.kind() == SyntaxKind::L_CURLY)
+            .map(|it| it.text_range())?;
```

Comment:
```suggestion
        let where_to_insert = trait_def
            .assoc_item_list()?
            .l_curly()?
            .text_range();
```

We also need to accomodate for macros here again though, as this range might come from within a macro expansion